source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-a-new-widget.rst

```diff
@@ -0,0 +1,199 @@
+
+A good widget allows the end-user to customize the widget to suit their needs. An example could be to allow the user to control the range of the number slider, that is, its maximum and minimum or the orientation of the slider itself. The view of the widget or how it looks is defined using FXML. ``FXML`` is an XML based language that is useful for defining the static layout of the widget (Panes, Labels and Controls).
+
+More about FXML can be found `here <https://openjfx.io/javadoc/11/javafx.fxml/javafx/fxml/doc-files/introduction_to_fxml.html>`_.
+
+Defining a Widget's FXML
+------------------------
+In this example, we will create two sliders to help us control the X and Y coordinates of our Point2D data type we created in previous sections. It is helpful to place the FXML file in the same package as the Java class.
+
+In order to create an empty, blank window for our widget, we need to create a ``Pane``. A Pane is a parent node that contains other child nodes, in this case, 2 sliders.
+There are many different types of Pane, they are as noted:
+
+- Stack Pane
+
+   - Stack Panes allow elements to be overlaid. Also, StackPanes by default center child nodes.
+
+- Grid Pane
+
+   - Grid Panes are extremely useful defining child elements using a coordinate system by creating a flexible grid of rows and columns on the pane.
+
+- Flow Pane
+
+   - Flow Panes wrap all child nodes at a boundary set. Child nodes can flow vertically (wrapped at the height boundary for the pane) or horizontally (wrapped at the width boundary of the pane).
+
+- Anchor Pane
+
+   - Anchor Panes allow child elements to be placed in the top, bottom, left side, right side, or center of the pane.
+
+Layout panes are also extremely useful for placing child nodes in one horizontal row using a `HBox <https://docs.oracle.com/javase/8/javafx/api/javafx/scene/layout/HBox.html>`_ or one vertical column using a `VBox <https://docs.oracle.com/javase/8/javafx/api/javafx/scene/layout/VBox.html>`_.
+
+The basic syntax for defining a Pane using FXML would be as the following:
+
+.. code-block:: xml
+
+   <?import javafx.scene.layout.*?>
+   <GridPane xmlns:fx="http://javafx.com/fxml/1" fx:controller="/path/to/widget/class" fx:id="root">
+      ...
+   </GridPane>
+
+Creating A Widget Class
+-----------------------
+
+Now that we have a Pane, we can now add child elements to that pane. In this example, we can add two slider objects. Remember to add an ``fx:id`` to each element so they can be reference in our Java class we will make later on. We will use a ``VBox`` to position our slider on top of each other.
+
+.. code-block:: xml
+
+   <?import javafx.scene.layout.*?>
+   <GridPane xmlns:fx="http://javafx.com/fxml/1" fx:controller="/path/to/widget/class" fx:id="root">
```

Comment:
Why GridPane here? StackPane would be better given that there's only one direct child node